conda/gateways/repodata/__init__.py

```diff
@@ -709,6 +714,251 @@ def timeout(self):
+            if context.use_index_cache:
+                log.debug(
+                    "Using cached repodata for %s at %s because use_cache=True",
+                    self.url_w_repodata_fn,
+                    self.cache_path_json,
+                )
+
+                _internal_state = self.read_cache()
+                return _internal_state
+
+            stale = cache.stale()
+            if (not stale or context.offline) and not self.url_w_subdir.startswith(
+                "file://"
+            ):
+                timeout = cache.timeout()
+                log.debug(
+                    "Using cached repodata for %s at %s. Timeout in %d sec",
+                    self.url_w_repodata_fn,
+                    self.cache_path_json,
+                    timeout,
+                )
+                _internal_state = self.read_cache()
+                return _internal_state
+
+            log.debug(
+                "Local cache timed out for %s at %s",
+                self.url_w_repodata_fn,
+                self.cache_path_json,
+            )
+
+        try:
+            try:
+                repo = self._repo
+                if hasattr(repo, "repodata_parsed"):
+                    raw_repodata = repo.repodata_parsed(cache.state)  # type: ignore
+                else:
+                    raw_repodata = repo.repodata(cache.state)  # type: ignore
+            except RepodataIsEmpty:
+                if self.repodata_fn != REPODATA_FN:
+                    raise  # is UnavailableInvalidChannel subclass
+                # the surrounding try/except/else will cache "{}"
+                raw_repodata = None
+            except RepodataOnDisk:
+                # used as a sentinel, not the raised exception object
+                raw_repodata = RepodataOnDisk
+
+        except Response304ContentUnchanged:
+            log.debug(
+                "304 NOT MODIFIED for '%s'. Updating mtime and loading from disk",
+                self.url_w_repodata_fn,
+            )
+            cache.refresh()
+            # touch(self.cache_path_json) # not anymore, or the a separate file is invalid
+            # self._save_state(mod_etag_headers)
+            _internal_state = self.read_cache()
+            return _internal_state
+        else:
+            try:
+                if raw_repodata is RepodataOnDisk:
+                    # this is handled very similar to a 304. Can the cases be merged?
+                    # we may need to read_bytes() and compare a hash to the state, instead.
+                    # XXX use self._repo_cache.load() or replace after passing temp path to jlap
+                    raw_repodata = self.cache_path_json.read_text()
+                    cache.state["size"] = len(raw_repodata)  # type: ignore
+                    stat = self.cache_path_json.stat()
+                    mtime_ns = stat.st_mtime_ns
+                    cache.state["mtime_ns"] = mtime_ns  # type: ignore
+                    cache.refresh()
```

Comment:
This cache.refresh() is redundant with this branch's version of jlap/fetch.py but is still needed with the "classic" fetcher. How should the fetcher communicate that the cache [metadata] is also up-to-date?

Reply:
Could you pass in which fetcher is currently in use, to decide whether to call `cache.refresh()` or not?

Either by requesting that information from conda's context, or by implementing a `ClassicRepodataCache` and a `JLAPRepodataCache` and the related `ClassicRepodataFetch` and `JLAPRepodataFetch` configured with them?

Reply:
We do already check with hasattr; maybe the contract of the new method is that the cache does not have to be updated on the caller side.

Reply:
That would suffice to me